development/extensions/tutorial_php_events_and_extending_extensions.rst

```diff
@@ -0,0 +1,305 @@
+  afterwards.
+
+Using service collections
+=========================
+In 3.1, a new concept is that of "services". You can read up on exactly what a
+service is `here <http://symfony.com/doc/current/book/service_container.html>`_;
+the rest of this guide will assume you have basic knowledge of services and how
+to use them (if you have further questions, feel free to ask). A service
+collection is basically what it sounds like: a collection of services. Basically,
+when you define your service, you give it a special "tag", which associates it
+with a collection of services. Later on, this can be used to easily get a
+list of services for use by your extension.
+
+To use this, you first create a service for your service collection. This will
+actually point to a core phpBB object called ``phpbb\di\service_collection``.
+
+.. code-block:: yaml
+
+  acme.demo.foobar_collection:
+      class: phpbb\di\service_collection
+      arguments:
+          - '@service_container'
+      tags:
+          - { name: service_collection, tag: acme.demo.foobar_service }
+
+Of course the service name and the tag name may be whatever you like.
+
+.. note::
+  To prevent duplicates in your tag you should use a unique tag name, preferably
+  in the form of acme.demo.tagname.
+
+Now, when you want to add a service to your collection, just do the following:
+
+.. code-block:: yaml
+
+  acme.demo.thing_one:
+      class: acme\demo\thingy\thing_one
+      tags:
+          - { name: acme.demo.foobar_service }
+
+Notice that the tag "name" value here corresponds to the tag "tag" value in the
+previous service definition. Also, keep in mind that if the class (in this case,
+``thing_one``) extends another class, you will need to provide the correct services
+and values for any parameters defined in the constructor for the parent class,
+if necessary.
+
+Finally, to use the collection of services, just pass the first service as an
+argument to another service class. For instance, let's say I have a manager
+object for my foobar extension and I want the manager to know about all of the
+services in the "foobar_collection" service. When defining the manager class, I
+just have to give the first service I showed you as an argument to the manager
+service.
+
+.. code-block:: yaml
+
+  acme.demo.foobar_manager:
+      arguments:
+          - '@acme.demo.foobar_collection'
+
+That argument will return an instance of ``phpbb\di\service_collection``, which
+extends ``ArrayObject`` so it can be used as an array containing the service name
+of each item in the collection as the key, and an instance of each of the items
+as the corresponding value.
+
+This system is used in the core for several features, including notifications
+and authentication providers.
+
+Ordered service collection
+--------------------------
+.. note::
+  Ordered service collections are introduced in phpBB 3.2.
+
+Ordered service collections are based on a normal service collection, but the
```

Comment:
I would start with something like this sentence too:
Ordered service collections allow you to define the order in which services are loaded, which is especially useful in cases where service priority and/or dependency requires they be loaded in a specified order.